core/src/data_sources/data_source.rs

```diff
@@ -730,25 +793,135 @@ impl DataSource {
             })
             .buffer_unordered(16)
             .map(|r| match r {
-                Err(e) => Err(anyhow!("DataSource document retrieval error: {}", e))?,
+                Err(e) => Err(anyhow!("datasource document retrieval error: {}", e))?,
                 Ok(r) => r,
             })
             .try_collect::<Vec<_>>()
             .await?;
 
-        // Merge the chunks with the documents.
-        let mut documents = documents
-            .into_iter()
+        let l_qdrant_client = Arc::new(Mutex::new(qdrant_client));
+        let mut documents = futures::stream::iter(documents)
             .map(|mut d| {
-                let chunks = chunks
+                let mut chunks = chunks
                     .iter()
                     .filter(|(document_id, _)| document_id == &d.document_id)
                     .map(|(_, c)| c.clone())
                     .collect::<Vec<Chunk>>();
-                d.chunks = chunks;
-                d
+                // look around a bit for arc/mutex stuff but I don't know if this is the right way to do this
```

Comment:
let's have a big `if !target_document_tokens.is_none()` here first so that we guarantee we don't do anything at all extra if it's not specified

Reply:
I should put the check inside the async process though? For simplicity

Reply:
No let's not do anything async if we don't need to. 1 check outside that just assigns d.chunks and returns d and async only if needed :+1: